src/cli.c

```diff
@@ -148,22 +149,70 @@ static void send_header(quicly_stream_t *stream, int is_http1, int status, const
-    char buf[1024];
-    size_t n;
+    static const quicly_streambuf_sendvec_callbacks_t send_file_callbacks = {flatten_file_vec, discard_file_vec};
+    int fd;
+    struct stat st;
 
-    if ((fp = fopen(fn, "rb")) == NULL)
+    if ((fd = open(fn, O_RDONLY)) == -1)
         return 0;
-    send_header(stream, is_http1, 200, mime_type);
-    while ((n = fread(buf, 1, sizeof(buf), fp)) != 0)
-        quicly_streambuf_egress_write(stream, buf, n);
-    fclose(fp);
+    if (fstat(fd, &st) != 0) {
+        close(fd);
+        return 0;
+    }
 
+    send_header(stream, is_http1, 200, mime_type);
+    quicly_sendbuf_vec_t vec = {&send_file_callbacks, (size_t)st.st_size, (void *)(intptr_t)fd};
+    quicly_streambuf_egress_write_vec(stream, &vec);
     return 1;
 }
 
+static int flatten_sized_text(quicly_sendbuf_vec_t *vec, void *dst, size_t off, size_t len)
+{
+    static const char pattern[] =
+        "hello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello world\nhello "
+        "world\nhello world\nhello world\nhello world\nhello world\nhello world\n";
+
+    assert(len < sizeof(pattern) - 13); /* pattern is bigger than MTU size */
```

Comment:
Why -13?

Reply:
Is this to allow for the offset to be worst case of 11?

Reply:
Correct. I think that the correct code should be `assert(len <= sizeof(pattern) - 12)` (note that there is an implicit `'\0'` exists at the end of the character array).

Though I think we can refactor the code for readability.

Reply:
Done in b9a308b.